lightpath/controller.py

```diff
@@ -42,115 +49,402 @@ def __init__(self, client, endstations=None):
         self.client = client
         self.containers = list()
         self.beamlines = dict()
+        self.graph = nx.DiGraph()
+        self.sources = set()
+
+        # initialize graph -> self.graph
+        self.load_facility()
+
         endstations = endstations or beamlines.keys()
-        # Find the requisite beamlines to reach our endstation
+        # # Find the requisite beamlines to reach our endstation
         for beamline in endstations:
             self.load_beamline(beamline)
 
-    def load_beamline(self, endstation):
+    def load_facility(self):
+        """
+        Load the facility from the provided happi client.
+
+        The facility is represented by a directed graph, with
+        devices as nodes.  Edge weights are initialized as 0, and
+        labeled with their branch.
+
+        The facility graph is created by combining subgraphs that
+        each contain all the devices on a given branch.
         """
-        Load a beamline from the provided happi client
+        results = self.client.search_range(key='z', start=0.0, end=math.inf,
+                                           active=True, lightpath=True)
```

Comment:
The main thing this has over a search for just `active=True, lightpath=True` is the enforcement that `z` exists and is non-negative, right? Some day we may consider making the "position" key configurable... Today is not that day.

Reply:
Yep, Those are the conditions for a device to be valid in lightpath's concept of a facility.  I thought about trying to make the "position" key more configurable, but the concept should be reasonably extensible to other light sources (synchrotrons just define z as distance ccw from injection or something?)